clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1520,6 +1523,9 @@ private Fetch<K, V> pollForFetches(Timer timer) {
             return fetch;
         }
 
+        // send any new fetches (won't resend pending fetches)
+        sendFetches(timer);
```

Comment:
We're not polling for the incoming responses in `sendFetches()`, just enqueuing the outgoing requests. This mimics the `ClassicKafkaConsumer` in that the requests are enqueued in its `sendFetches()` but then toward the bottom of `pollForFetches()` `client.poll()` is invoked to wait for the results of the fetch requests.

Reply:
well the `sendFetches` blocks until the `CreateFetchRequestsEvent` completes, and that only happens on fetchMgr.poll
https://github.com/apache/kafka/blob/aae7e97fa041dac34a0788f90e837ff7f5ec3846/clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java#L115
So when the `sendFetches` completes we did poll the manager right? (and depending on time, maybe we did poll the client.poll too, which happens in the background right after polling all managers). That's why the log for "Polling for fetches" made sense to me before the `sendFetches`, but am I missing another poll happening after the log line maybe? (where it is now)

Reply:
The two `ConsumerDelegate` implementations work differently:

* `AsyncKafkaConsumer`: `FetchRequestManager.poll()` will complete the event's `Future` on the background thread before it exits, i.e. before the thread starts the network I/O. Completing the `Future` starts the application thread racing toward logging that message and the background thread racing toward starting network I/O. I'll admit—I haven't dug through the code to surmise the relative costs of each thread's work before either cross their respective finish lines to win.
* `ClassicKafkaConsumer`: `Fetcher.sendFetchesInternal()` calls `ConsumerNetworkClient.send()` to enqueue the request, but then it calls `NetworkClient.wakeup()`. Since the same `ConsumerNetworkClient` instance used by the consumer is also used by `AbstractCoordinator.HeartbeatThread`, it's technically possible that the heartbeat thread's `run()` method could start network I/O when it calls `NetworkClient.pollNoWakeup()`. Granted, that's a race that the application thread is much more likely to win given that the heartbeat thread runs much less frequently.

Here are some points to consider:

* The definition of the term "poll" as used in the log is open to interpretation. The term "poll" is everywhere, making its meaning ambiguous at any given point of use 😢
* I agree there is a race condition (for both consumers, but more likely for the new consumer) that could result the the log message being emitted after the network I/O has commenced
* For this to pose a problem to users, there needs to be other log entries that we're racing with, right?. We're trying to avoid the condition where the user is confused/mislead because the entries in the log are emitted in non-deterministic ordering.
* The log line in question is only output at level `TRACE`, which I _assume_ is very rare for users to enable.

Given the above, I'm of the opinion that it's an exercise in hair splitting to alter the logging. However, I could also just change it which would have been way less effort than researching, thinking, and composing this response 🤣

If we leave the log line as it is, what would the effect be for the user?

Reply:
I surely didn't intend for you to put up that long response he he, sorry. It's not about the log line per-se, it's about the alignment on where the poll happens. The classic consumer logs "Polling for records", then calls `client.poll`. vs us here we do `sendFetches` (which triggers the client.poll async in the background thread because it blocks until we poll the fetch manager), then log "Polling for fetches...". 

That's the diff I saw and just wanted to understand/align on where the poll happens: once we trigger `sendFetches` (blocking), the client.poll will happen in the background anytime, not controlled by the app thread. Agreed? If so I'm ok with leaving the log unchanged, understanding it could come out after the client.poll happened.

Reply:
> That's the diff I saw and just wanted to understand/align on where the poll happens: once we trigger sendFetches (blocking), the client.poll will happen in the background anytime, not controlled by the app thread. Agreed?

Agreed—the background thread is going to move from calling each of the `RequestManager`’s `poll()` method to `NetworkClient.poll()` method without the intervention of the application thread.

> If so I'm ok with leaving the log unchanged, understanding it could come out after the client.poll happened.

Thanks!